rcl_yaml_param_parser/src/node_params.c

```diff
@@ -116,30 +133,33 @@ void rcl_yaml_node_params_fini(
     return;
   }
 
-  if (NULL != node_params_st->parameter_names) {
-    for (size_t parameter_idx = 0U; parameter_idx < node_params_st->num_params;
-      parameter_idx++)
-    {
-      char * param_name = node_params_st->parameter_names[parameter_idx];
-      if (NULL != param_name) {
-        allocator.deallocate(param_name, allocator.state);
-      }
-    }
-    allocator.deallocate(node_params_st->parameter_names, allocator.state);
-    node_params_st->parameter_names = NULL;
-  }
-
-  if (NULL != node_params_st->parameter_values) {
-    for (size_t parameter_idx = 0U; parameter_idx < node_params_st->num_params;
-      parameter_idx++)
-    {
-      rcl_yaml_variant_fini(&(node_params_st->parameter_values[parameter_idx]), allocator);
+  char * param_name = NULL;
+  rcl_variant_t * param_value = NULL;
+  rcutils_ret_t ret = rcutils_hash_map_get_next_key_and_data(
+    &node_params_st->node_params_map, NULL, &param_name, &param_value);
+  while (RCUTILS_RET_OK == ret) {
+    rcl_yaml_variant_fini(param_value, allocator);
+    ret = rcutils_hash_map_unset(&node_params_st->node_params_map, &param_name);
+    if (RCUTILS_RET_OK != ret) {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Failed to unset key '%s' for hash map of a node parameter", param_name);
+      return;
     }
 
-    allocator.deallocate(node_params_st->parameter_values, allocator.state);
-    node_params_st->parameter_values = NULL;
+    allocator.deallocate(param_name, allocator.state);
+    param_name = NULL;
+    allocator.deallocate(param_value, allocator.state);
+    param_value = NULL;
+    ret = rcutils_hash_map_get_next_key_and_data(
+      &node_params_st->node_params_map, NULL, &param_name, &param_value);
+  }
+  if (RCUTILS_RET_HASH_MAP_NO_MORE_ENTRIES != ret) {
+    RCUTILS_SET_ERROR_MSG("Failed to get next item for node parameters");
```

Comment:
error message is not exactly correct.

```suggestion
    RCUTILS_SET_ERROR_MSG("Failed to remove all entries in the hash map");
```